### PR TITLE
feature: Hook for checking cache status & reloadOnDowngrade

### LINKED ISSUE #17: [feature request] Hook for requesting a version check
Hi! Thanks for this simple but useful package.

Could we do a simple hook that returned a function which we could call to trigger a version check?

Currently when adding this wrapper to our (react-admin) SPA, user still has to reload the page for react-cache-buster to do its thing.

If we had a way to trigger the version check whenever, we could run it on each route change or something like that, and make sure our users (which tend to leave our app open for days) always get the freshest version.

I could try to take a crack at it though i'm not sure when i'll have the time.

Thanks!

### PR DESCRIPTION
**Pull request by MartinCura**
Closes #17 

This PR makes CacheBuster into a Provider with a Context so that any distant child component can call `checkCacheStatus()` without needing the user to reload the page, effectively allowing them to maybe bust the cache whenever they want.

[Edit] I've also now added a new prop `reloadOnDowgrade`, which i think is quite useful.

Other small changes, which can be removed if you want:
* add `@ts-check` to top of file, so IDE tells on type erros without completely switching to TS
* remove arg from `window.location.reload()` as it was a type error (it has no arguments: https://html.spec.whatwg.org/multipage/nav-history-apis.html#dom-location-reload)
* wrap `checkCacheStatus` in a `useCallback()`
* updated `caniuse` in lockfile

**Please tell me if there's anything i can help you with to merge this. :)**

Example usage of new hook:

```jsx
// ...
import { useLocation } from 'react-router-dom';
import CacheBuster, { useCacheBuster } from 'react-cache-buster';

const isEnabled = process.NODE_ENV === 'production';

const AppUpdater = ({ children }: { children: ReactNode }) => {
    const VersionChecker = ({ children }: { children: ReactNode }) => {
        const [skippedFirstRender, setSkippedFirstRender] =
            useState<boolean>(false);
        const location = useLocation();
        const { checkCacheStatus } = useCacheBuster();

        // Every time the location changes, check for a new version
        // as this (for some apps) is an unobtrusive moment for this.
        useEffect(() => {
            if (!isEnabled) {
                return;
            }
            if (!skippedFirstRender) {
                setSkippedFirstRender(true);
                return;
            }
            checkCacheStatus();
        }, [location, checkCacheStatus]);

        return <>{children}</>;
    };

    return (
        <CacheBuster
            currentVersion={packageInfo.version}
            isEnabled={isEnabled}
            isVerboseMode={false}
            loadingComponent={<p style={{ margin: '20px' }}>Loading app...</p>}
            // really make sure browser doesn't cache meta.json
            metaFileDirectory={`/meta.json?t=${Date.now()}#`}
        >
            <VersionChecker>{children}</VersionChecker>
        </CacheBuster>
    );
};
```

### CODE REVIEW

**Comment by MartinCura** [comment]
Added new `reloadOnDowngrade` prop.

**Comment by CagriAldemir** [comment]
Hi @MartinCura. Thank you for your development. I think it's a good idea too. I will review and publish the new version this week.

**Comment by robmadden** [comment]
Hi there, thanks for creating this library.

I know it's only been one day, but is there any chance you could release this feature? I was hoping to use this library in one of my projects, thanks!

File: README.md [hunk 1]
```diff
@@ -60,7 +60,7 @@ const App = () => {
       metaFileDirectory={'.'} //If public assets are hosted somewhere other than root on your server.
     >
 
-      //Your actual root component...
+      {/* // Your actual root component... */}
```
Comment (CagriAldemir) [inline README.md]: Can you revert this comment line to `// Your actual root component...` ?

File: README.md [hunk 2]
```diff
@@ -89,7 +89,8 @@ The process works as follows;
...
-| onCacheClear      | Function | :x:                | This function is called before clearing the cache when a new version is found. If you pass this prop, cache clearing is not performed. Instead, the cache clearing function is sent as a parameter to this function and **you are expected to call this function**.<br/>Structure of the function: `oncacheClear: (refreshCacheAndReload: () => Promise<void>) => void;` |
+| reloadOnDowngrade | Boolean  | :x:                | Whether to also bust the cache and reload if the version fetched from the server is lower.                                                                                                                                                                                                                                                                               |
```
Comment (CagriAldemir) [inline README.md]: Can you add `Default: **false**` to end of description?

File: src/CacheBuster.js [hunk 3]
```diff
@@ -76,7 +89,7 @@ function CacheBuster({
         await Promise.all(cacheDeletionPromises);
 
         log('The cache has been deleted.');
-        window.location.reload(true);
+        window.location.reload();
```
Comment (CagriAldemir) [inline src/CacheBuster.js]: The "forceRefresh" parameter is still [supported by the Firefox browser](https://developer.mozilla.org/en-US/docs/Web/API/Location/reload#parameters). So you better add the parameter again.

**Review by CagriAldemir — CHANGES_REQUESTED** [review]
I added some comments. Can you check them?

You also added a hook. Thank you for that. However, I would be very happy if you add the details about this hook to the README file and add a simple example of its use in the example project located in [this directory](https://github.com/CagriAldemir/react-cache-buster/tree/master/example).

File: src/CacheBuster.js (hunk 3 above)
Comment (MartinCura) [reply src/CacheBuster.js]: Missed that! Thanks for pointing it out.

**Comment by MartinCura** [comment]
There you go, @CagriAldemir. Happy to change anything else as you wish, but also feel free to fix any nits yourself (i think you should be able to push to my branch) 😃 

We're already using this so very happy if we can merge it soon.

**Comment by robmadden** [comment]
Thanks for your efforts!

**Comment by robmadden** [comment]
Friendly ping: any update here?

**Comment by MartinCura** [comment]
> I added some comments. Can you check them?
> 
> You also added a hook. Thank you for that. However, I would be very happy if you add the details about this hook to the README file and add a simple example of its use in the example project located in [this directory](https://github.com/CagriAldemir/react-cache-buster/tree/master/example).

One more gentle ping, sorry 😄 I made the small changes you asked for. Is there anything else i can do to get this merged? 😇 Thank you.

**Comment by robmadden** [comment]
👋  any update here?

**Review by CagriAldemir — APPROVED** [review]

**Comment by CagriAldemir** [comment]
@MartinCura @robmadden Hello there. I could not review due to my workload, but I have just published the new version of Cache Buster. Thanks for support.

**Comment by robmadden** [comment]
@CagriAldemir no worries, thanks for following up!

**Comment by MartinCura** [comment]
No worries, thank you so much ! Already updating this dep in our repo.

**Pull request merged**